### PR TITLE
[SPARK-48935][SQL][TESTS] Make `checkEvaluation` directly check the `Collation` expression itself in UT

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to:
- make `checkEvaluation` directly check the `Collation` expression itself in UT, rather than `Collation(...).replacement`.
- fix an `miss` check in UT.

### Why are the changes needed?
When checking the `RuntimeReplaceable` expression in UT, there is no need to write as `checkEvaluation(Collation(Literal("abc")).replacement, "UTF8_BINARY")`, because it has already undergone a similar replacement internally.
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala#L75

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Update existed UT.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollationRegexpExpressionsSuite.scala [hunk 1]
```diff
@@ -120,50 +121,84 @@ class CollationRegexpExpressionsSuite extends SparkFunSuite with ExpressionEvalH
   }
 
   test("MultiLikeBase regexp expressions with collated strings") {
-    val nullStr = Literal.create(null, StringType)
-    // Supported collations (StringTypeBinaryLcase)
-    val binaryCollation = StringType(CollationFactory.collationNameToId("UTF8_BINARY"))
-    val lowercaseCollation = StringType(CollationFactory.collationNameToId("UTF8_LCASE"))
     // LikeAll
```
Comment (panbingkun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollationRegexpExpressionsSuite.scala:124]: Although the following modifications are not related to this PR, they are only made to maintain consistency with the rest of the checks in this PR.

**Comment by panbingkun** [comment]
cc @cloud-fan

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 2]
```diff
@@ -88,7 +88,10 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
  */
 @Stable
 case object StringType extends StringType(0) {
-  private[spark] def apply(collationId: Int): StringType = new StringType(collationId)
+  private[spark] def apply(collationId: Int): StringType = {
+    assert (collationId >= 0 && collationId <= (1 << 12))
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:92]: where did we check collation id before?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollationExpressionSuite.scala [hunk 3]
```diff
@@ -67,16 +67,16 @@ class CollationExpressionSuite extends SparkFunSuite with ExpressionEvalHelper {
   }
 
   test("collation on non-explicit default collation") {
-    checkEvaluation(Collation(Literal("abc")).replacement, "UTF8_BINARY")
+    checkEvaluation(Collation(Literal("abc")), "UTF8_BINARY")
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollationExpressionSuite.scala:70]: does `checkEvaluation`  take care of `RuntimeReplaceable`?

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala (hunk 2 above)
Comment (panbingkun) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala]: It did not check `collationId`, it only checked `collationName`,
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala#L94
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java#L391
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java#L615

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollationExpressionSuite.scala (hunk 3 above)
Comment (panbingkun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollationExpressionSuite.scala:70]: Yea, as follows:
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala#L87-L90

https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala#L79-L82

recursion
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala#L74-L75

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala (hunk 2 above)
Comment (panbingkun) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala]: Additionally, exceptions may be thrown below due to invalid `collationId`, but unfortunately, they are all `functions `that only `trigger` when `called`
https://github.com/apache/spark/blob/83a97917d46debe9c3ebb24cf2540b07f6a902ec/sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala#L32-L46

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 4]
```diff
@@ -88,7 +88,10 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
  */
 @Stable
 case object StringType extends StringType(0) {
-  private[spark] def apply(collationId: Int): StringType = new StringType(collationId)
+  private[spark] def apply(collationId: Int): StringType = {
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:91]: I checked several callers of this function. The input collation id is mostly calculated from collation name. This assertion doesn't seem to be necessary and it's not cheap. Shall we revert?

I'm fine with other cleanups in this PR.

Comment (panbingkun) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala]: Okay, I have revert it. 
Thank you!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by uros-db** [comment]
hi folks @panbingkun @cloud-fan, just looking at this now - are we sure that we want to call "checkEvaluation" on `Collation(Literal("abc"))` (instead of `Collation(Literal("abc")).replacement`)?

afaik, we can't call `.eval `on `RuntimeReplaceable`s

`trait RuntimeReplaceable extends Expression` (Expression.scala):
```
  final override def eval(input: InternalRow = null): Any =
    throw QueryExecutionErrors.cannotEvaluateExpressionError(this)
```

**Pull request closed without merging**